sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/BatchGroupUtils.scala

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.execution.python
+
+import scala.collection.mutable
+
+import ai.rapids.cudf
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+
+/**
+ * Basic functionality to deal with groups in a batch.
+ *
+ * It plays the similar role to the Spark 'PandasGroupUtils', but dealing with 'ColumnarBatch',
+ * instead of 'InternalRow'.
+ */
+private[python] object BatchGroupUtils extends Arm {
+
+  /**
+   * Returns the deduplicated attributes of the Spark plan, and the argument offsets of the
+   * keys and values.
+   *
+   * The deduplicated attributes are needed because the Spark plan may contain an attribute
+   * twice; once in the key and once in the value. For any such attribute we need to
+   * deduplicate.
+   *
+   * The argument offsets are used to distinguish grouping attributes and data attributes
+   * as following:
+   *
+   * argOffsets[0] is the length of the argOffsets array
+   * argOffsets[1] is the length of grouping attribute
+   * argOffsets[2 .. argOffsets[1]+2] is the argument offsets for grouping attributes
+   * argOffsets[argOffsets[1]+2 .. ] is the argument offsets for data attributes
+   */
+  def resolveArgOffsets(
```

Comment:
Thanks for review and updated.

In the latest change, I dropped the comment mostly copied from Spark, and try to explain as much as possible per my understanding. Hope this can help.